File: app/src/main/java/com/example/djikstra/Djikstra.java
package com.example.djikstra;

import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.List;

public class Djikstra {
    public void calculatedShortestPath(Node source){
        source.setDistance(0);
        Set<Node> settledNodes = new HashSet<>();
        Queue<Node> unsettledNodes = new PriorityQueue<>(Collections.singleton(source));
        while (!unsettledNodes.isEmpty()) {
            Node currentNode = unsettledNodes.poll();
            currentNode.getAdjacentNodes()
                    .entrySet().stream()
                    .filter(entry -> !settledNodes.contains(entry.getKey()))
                    .forEach(entry -> {
                            evaluateDistanceAndPath(entry.getKey(), entry.getValue(), currentNode);
                            unsettledNodes.add(entry.getKey());
                    });
            settledNodes.add(currentNode);
        }
    }
    private void evaluateDistanceAndPath(Node adjacentNode, Integer edgeWeight, Node sourceNode) {
        Integer newDistance = sourceNode.getDistance() + edgeWeight;
        if (newDistance < adjacentNode.getDistance()){
            adjacentNode.setDistance(newDistance);
            adjacentNode.setShortestPath(Stream.concat(sourceNode.getShortestPath().stream(), Stream.of(sourceNode)).collect(Collectors.toList())
            );
        }
    }
}

File: app/src/main/java/com/example/djikstra/djikstra.kt
package com.example.djikstra

import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

class djikstra {
    fun calculatedShortestPath(source: node) {
        source
        val settledNodes: MutableSet<Node> = HashSet()
        val unsettledNodes: Queue<Node> = PriorityQueue(setOf(source))
        while (!unsettledNodes.isEmpty()) {
            val currentNode = unsettledNodes.poll()
            currentNode.getAdjacentNodes()
                .entrySet().stream()
                .filter { entry -> !settledNodes.contains(entry.getKey()) }
                .forEach { entry ->
                    evaluateDistanceAndPath(entry.getKey(), entry.getValue(), currentNode)
                    unsettledNodes.add(entry.getKey())
                }
            settledNodes.add(currentNode)
        }
    }

    private fun evaluateDistanceAndPath(adjacentNode: Node, edgeWeight: Int, sourceNode: Node) {
        val newDistance: Int = sourceNode.getDistance()+ edgeWeight
        if (newDistance < adjacentNode.getDistance()) {
            adjacentNode.setDistance(newDistance)
            adjacentNode.setShortestPath(
                Stream.concat(sourceNode.getShortestPath().stream(), Stream.of(sourceNode)).collect(
                    Collectors.toList()
                )
            )
        }
    }
}